### PR TITLE
Added You.i Engine support

### PR DESCRIPTION
**Pull request by simongranger**
## Proposed changes

Adding support to You.i Engine into Appium
http://www.youi.tv/youi-engine/
## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._
- [X] I have read the [CONTRIBUTING](/CONTRIBUTING.md) doc
- [X] I have signed the [CLA](https://docs.google.com/forms/d/1lOfXRw_0VCk7gYzjj4WLetGu7yelDVo5LWh0z7pGftE/viewform)
- [X] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)
- [X] Any dependent changes have been merged and published in downstream modules
## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by moizjv** [comment]
Can one of the admins verify this patch?

**Comment by simongranger** [comment]
'npm test' works fine over here

**Comment by imurchie** [comment]
It's weird that the `appium-youi-driver` package isn't getting installed in Node 4 and 5. Does this work locally?

**Comment by simongranger** [comment]
Yes,
It works with 5 but not 4 or 0.12
I have only tried with 5 locally.
I first need to understand how to install the older versions...
:S

**Comment by SrinivasanTarget** [comment]
Try older versions of node using node-version-manager, npm install -g n

**Comment by simongranger** [comment]
I get the following error:

$ nvm install 4

/Users/simongranger/.nvm/node-v4 doesn't exist, (need to downloaded first?)
4

I installed node using brew. Can brew and nvm coexist?

**Comment by imurchie** [comment]
Hmm. nvm ought to take over, and manage things.

What do you get from `nvm ls`?

**Comment by simongranger** [comment]
I found _nodeVersion and _npmVersion get added to the package.json when
downloaded from npm.
They are not in our github repo.

Could we have forced a minimum version when publishing to npm?

  "_from": "appium-youi-driver@latest",
  "_id": "appium-youi-driver@1.0.9",
  "_inCache": true,
  "_installable": true,
  "_location": "/appium-youi-driver",
  "_nodeVersion": "5.5.0",
  "_npmOperationalInternal": {
    "host": "packages-12-west.internal.npmjs.com",
    "tmp":
"tmp/appium-youi-driver-1.0.9.tgz_1456860453207_0.9975910184439272"
  },
  "_npmUser": {
    "email": "andrew@youi.tv",
    "name": "aemmons"
  },
  "_npmVersion": "3.5.3",

**Comment by imurchie** [comment]
I like to publish with the minimum version we support (0.12 for Appium). I'm not sure if it causes problems.

**Comment by simongranger** [comment]
$ nvm ls

"local" not implemented yet

**Comment by imurchie** [comment]
Hmm. nvm seems broken. Can you delete your `~/.nvm` directory, then reinstall it?

**Comment by simongranger** [comment]
~/.nvm was never created
I only see ~/.npm

$ npm install -g nvm

npm WARN deprecated nvm@0.0.3: This is NOT the correct nvm. Visit
http://nvm.sh and use the curl command to install it.

/Users/simongranger/.node/bin/nvm ->
/Users/simongranger/.node/lib/node_modules/nvm/bin/nvm

/Users/simongranger/.node/lib

└─┬ nvm@0.0.3

  └── mkdirp@0.3.5

**Comment by imurchie** [comment]
Hmm. Ok. Can you install through https://github.com/creationix/nvm

**Comment by simongranger** [comment]
OK,
I got nvm installed properly.
I can see the same issue.
appium-youi-driver does not get installed with 'npm install' for older node
versions.

What is the proper way to publish in npm?
We have node versions in our npm copy while appium-ios-driver doesn't.
We were simply running 'npm install' followed by 'npm publish'

**Comment by imurchie** [comment]
Generally speaking, I publish with the lowest version of node that I want to support. In Appium's case, at the moment, that is 0.12. I then go through these steps

```
rm -rf node_modules
npm install --production
npm publish
```

**Comment by simongranger** [comment]
OK, so there we have it! We were publishing with node 5
I will republish shortly

**Comment by simongranger** [comment]
Publishing with an older version of node doesn't solve it.

**Comment by imurchie** [comment]
Oh! I think I figured out the problem. 

I guess node 0.12 doesn't respect `npm shrinkwrap`. Anyway, we use a system to freeze dependencies, called "npm shrinkwrap", which creates the file `npm-shrinkwrap.json` which completely specifies the version of each package. This needs to be updated to include your package.

I'm not entirely sure of the best process for updating it, since usually it involves running the full test suite and making sure everything is up-to-date and working. This is usually only done at release time.

**Comment by simongranger** [comment]
Is this something I can test locally?

**Comment by imurchie** [comment]
Alright. Sorry for the confusion and the delay. I've convinced everyone and removed `npm-shrinkwrap.json` from the `master` branch. Can you rebase your branch from master, and repush?

**Comment by simongranger** [comment]
OK, that seems to work locally. I can see 'appium-youi-driver' after 'npm
install'

I still get an error running 'npm test' with all versions of node (see
below)
Any idea why?

[11:10:19] Starting 'unit-test'...

[11:10:19] TypeError in plugin 'gulp-mocha'

Message:

    _lodash2.default.pairs is not a function

Stack:

TypeError: _lodash2.default.pairs is not a function

    at Object.<anonymous> (lib/driver.js:250:25)

    at Module._compile (module.js:413:34)

    at Object.Module._extensions..js (module.js:422:10)

    at Module.load (module.js:357:32)

    at Function.Module._load (module.js:314:12)

    at Module.require (module.js:367:17)

    at require (internal/module.js:16:19)

    at Object.<anonymous>
(/Users/simongranger/git/Appium1.5Stock/appium/node_modules/appium-youi-driver/build/index.js:7:18)

    at Module._compile (module.js:413:34)

    at Object.Module._extensions..js (module.js:422:10)

    at Module.load (module.js:357:32)

    at Function.Module._load (module.js:314:12)

    at Module.require (module.js:367:17)

    at require (internal/module.js:16:19)

    at Object.<anonymous> (lib/appium.js:9:60)

    at Module._compile (module.js:413:34)

    at Object.Module._extensions..js (module.js:422:10)

    at Module.load (module.js:357:32)

    at Function.Module._load (module.js:314:12)

    at Module.require (module.js:367:17)

    at require (internal/module.js:16:19)

    at Object.<anonymous> (lib/main.js:13:39)

    at Module._compile (module.js:413:34)

    at Object.Module._extensions..js (module.js:422:10)

    at Module.load (module.js:357:32)

    at Function.Module._load (module.js:314:12)

    at Module.require (module.js:367:17)

    at require (internal/module.js:16:19)

    at Object.<anonymous> (test/driver-e2e-specs.js:7:20)

    at Module._compile (module.js:413:34)

    at Object.Module._extensions..js (module.js:422:10)

    at Module.load (module.js:357:32)

    at Function.Module._load (module.js:314:12)

    at Module.require (module.js:367:17)

[11:10:19] 'unit-test' errored after 955 ms

On Thu, Mar 10, 2016 at 10:43 AM, Isaac A. Murchie <notifications@github.com

**Comment by imurchie** [comment]
Hmmm. It seems like maybe lodash 4 is being installed instead of 3? I am installing locally and looking.

**Comment by simongranger** [comment]
Yeah, I changed some versions in the appium-youi-driver to match the latest
available...
I will revert that change

On Thu, Mar 10, 2016 at 11:23 AM, Isaac A. Murchie <notifications@github.com

**Comment by imurchie** [comment]
Ok. Yes, lodahs 4 changed a bunch of functions. Including `_.pairs` became `_.toPairs`. Yay!

**Comment by simongranger** [comment]
Can you try running npm test again?

I now get:

node stable: PASS
node 0.12: PASS
node 4: FAIL

> node ./bin/install.js


Error: Cannot find module '../modules/web.dom.iterable'

    at Function.Module._resolveFilename (module.js:325:15)

    at Function.Module._load (module.js:276:25)

    at Module.require (module.js:353:17)

    at require (internal/module.js:12:17)

    at Object.<anonymous>
(/Users/simongranger/git/Appium1.5Stock/appium/node_modules/asyncbox/node_modules/babel-runtime/node_modules/core-js/library/fn/is-iterable.js:1:1)

    at Module._compile (module.js:409:26)

    at Object.Module._extensions..js (module.js:416:10)

    at Module.load (module.js:343:32)

    at Function.Module._load (module.js:300:12)

    at Module.require (module.js:353:17)

-

> fsevents@1.0.8 install
/Users/simongranger/git/Appium1.5Stock/appium/node_modules/fsevents

> node-pre-gyp install --fallback-to-build


[fsevents] Success:
"/Users/simongranger/git/Appium1.5Stock/appium/node_modules/fsevents/lib/binding/Release/node-v46-darwin-x64/fse.node"
is installed via remote

npm ERR! Darwin 15.3.0

npm ERR! argv "/Users/simongranger/.nvm/versions/node/v4.4.0/bin/node"
"/Users/simongranger/.nvm/versions/node/v4.4.0/bin/npm" "install"

npm ERR! node v4.4.0

npm ERR! npm  v2.14.20

npm ERR! code ELIFECYCLE


npm ERR! appium-selendroid-driver@1.2.3 install: `node ./bin/install.js`

npm ERR! Exit status 1

npm ERR!

npm ERR! Failed at the appium-selendroid-driver@1.2.3 install script 'node
./bin/install.js'.

npm ERR! This is most likely a problem with the appium-selendroid-driver
package,

npm ERR! not with npm itself.

npm ERR! Tell the author that this fails on your system:

npm ERR!     node ./bin/install.js

npm ERR! You can get information on how to open an issue for this project
with:

npm ERR!     npm bugs appium-selendroid-driver

npm ERR! Or if that isn't available, you can get their info via:

npm ERR!

npm ERR!     npm owner ls appium-selendroid-driver

npm ERR! There is likely additional logging output above.


npm ERR! Please include the following file with any support request:

npm ERR!     /Users/simongranger/git/Appium1.5Stock/appium/npm-debug.log





On Thu, Mar 10, 2016 at 11:28 AM, Isaac A. Murchie <notifications@github.com

**Comment by simongranger** [comment]
Can you try again?

**Comment by simongranger** [comment]
I still get an error with node v4.4.0

**Comment by simongranger** [comment]
FYI, I've updated appium-youi-driver to have appium-ios-driver as a
dependencie in package.json

**Comment by imurchie** [comment]
For me everything is a-ok!

* Node 0.12.10 with npm 2.14.12: ✔︎
* Node 4.3.1. with npm 3.8.1: ✔︎
* Node 4.4.0 with npm 3.8.1: ✔︎
* Node 5.8.0 with npm 3.7.3: ✔︎

**Comment by simongranger** [comment]
Great!
What's the next step?

**Comment by imurchie** [comment]
As it stands this needs to be updated to require at least a working version of the driver. `appium-youi-driver@1.0.13` is the latest on npm and ought to fix the Travis build (on that note, it would be preferable for the version of GitHub and that on npm were the same).

In order for the driver's inclusion in Appium not to be completely out of the blue, it would also be good to update the README's [Supported Platforms](https://github.com/appium/appium/blob/master/README.md#supported-platforms) page, and the [Platform Support](https://github.com/appium/appium/blob/master/docs/en/appium-setup/platform-support.md) page. This should include some information about what the system is and how to get documentation (of which I can find none).

That would constitute something that could be merged into the Appium codebase.

**Comment by triager** [comment]
Can one of the admins verify this patch?

**Comment by simongranger** [comment]
We have replaced youi with YouiEngine

**Comment by jlipps** [comment]
@simongranger looks good!

> In order for the driver's inclusion in Appium not to be completely out of the blue, it would also be good to update the README's Supported Platforms page, and the Platform Support page. This should include some information about what the system is and how to get documentation (of which I can find none).

> That would constitute something that could be merged into the Appium codebase.

Can we get a bit of documentation and links to the driver, setup instructions, etc...?

**Comment by simongranger** [comment]
Yes,
we'll look into that shortly

**Comment by imurchie** [comment]
I think the driver needs to be open sourced, as well. I could be wrong but I don't think we want to be depending on closed code. @jlipps ?

**Comment by jlipps** [comment]
Ah, I assumed it was. Yeah... We should not have a restrictively-licensed NPM package in our dependencies. I'm open to creating some kind of CLI command that downloads the package after a user confirmation, something like this:

```
appium --add-proprietary-driver "youi"
```

This could in essence just do a local `npm install`. The code in Appium itself could simply check its `node_modules` for the package and error out with instructions if it doesn't exist. That way we're not including closed source packages in our dependency tree.

But @simongranger, I'd definitely expect the `appium-youiengine-driver` code to be free and open source, even if at the end of the day there is some closed source binary it has to download on install in order to work. Does that make sense in your architecture?

**Comment by simongranger** [comment]
Would that still allow our customers to use our driver on AWS for examples?

**Comment by simongranger** [comment]
I think we are fine in having appium-youiengine-driver open source. We just
have to change the privacy setting. I think this is our first open source
collaboration so they are set to private by default.
Isn't all the code already exposed when you install the driver?

**Comment by jlipps** [comment]
@simongranger it's about much more than the code being exposed. Appium is released under a certain open source license which allows people a lot of freedom in embedding Appium or developing it. As such Appium cannot ship closed source or restrictive licensed dependencies otherwise it legally jeopardizes Appium itself. That is why we need to have Appium's dependencies be open source with an official open source license (we recommend Apache 2).

There is also the aspect of collaboration; it's better if as much as possible is open source and up on a site like GitHub so that the community can collaborate on it.

Appium can _interoperate_ with closed source tools, however. The cleanest way to do this is the same way we work with Apple's tools, for example: we do not bundle Apple software in Appium. Rather, we assume/require that the user has the appropriate Apple tools on their system in order for Appium's iOS automation to work. In other words, let's say there is some closed source youi.tv binary that the appium-youiengine-driver calls out to. The appium-youiengine-driver code should simply double-check that the binary is correctly installed, etc..., and then proceed on its way. You can then add documentation to Appium explaining how a user would go about getting this closed-source aspect set up.

**Comment by philgreenyouilabs** [comment]
@jlipps can we (You i Labs Inc.) retain copyright in the license? I see most of the licenses have:
`Copyright 2012-2016 Sauce Labs, Inc.`
Whereas some (e.g. https://github.com/appium/node-adb-client/blob/master/LICENSE) have:
`Copyright 2012-2013 Appium Committers`

**Comment by jlipps** [comment]
@philgreenyouilabs sure, in your own driver I don't see any reason you couldn't maintain the copyright. The license is what's more important.

**Comment by philgreenyouilabs** [comment]
@jlipps please see https://github.com/YOU-i-Labs/appium-youiengine-driver

**Comment by philgreenyouilabs** [comment]
Are there any more review comments or things that we need to address before we can merge this (other than the resolving the current conflicts that have arisen since this has been open)? thanks!

**Comment by jlipps** [comment]
@philgreenyouilabs I still haven't had time to review the code, but I did just create this issue:

https://github.com/YOU-i-Labs/appium-youiengine-driver/issues/21

Once that's resolved on the youiengine side it will also mean a change in this PR

**Comment by imurchie** [comment]
A couple of first thoughts. For the driver to be installable from source there needs to be a local copy of `gulp` (usually this is made a `devDependency` in `package.json`.

I would also recommend running lint on it. Appium's gulp plugins (which would be better installed as a dev dependency, it is currently a regular one) supports `gulp jshint` right out of the box, which can catch some potential issues (particularly things like using `==` instead of `===`) and extra defined variables.

To configure jshint you can copy https://github.com/appium/appium-ios-driver/blob/master/.jshintrc

I would also recommend `eslint`, which catches a lot more. We have a config for Appium here: https://github.com/appium/eslint-config-appium

**Comment by philgreenyouilabs** [comment]
@jlipps replied to your issue: https://github.com/YOU-i-Labs/appium-youiengine-driver/issues/21#issuecomment-228838884

File: lib/appium.js [hunk 1]
```diff
@@ -48,6 +49,9 @@ class AppiumDriver extends BaseDriver {
       } else if (caps.automationName.toLowerCase() === 'xcuitest') {
         // but if we do and it is 'XCUITest', act on it
         return WebDriverAgentDriver;
+      } else if (caps.automationName.toLowerCase() === 'youiengine') {
+        // but if we do and it is 'XCUITest', act on it
```
Comment (jlipps) [inline lib/appium.js]: this comment should be fixed

**Comment by jlipps** [comment]
This LGTM for now. I would still love to see something in the docs explaining what Youi is and how to get the system set up appropriately for using it (even if it is only saying there are no requirements beyond choosing a Youi-based app)

**Comment by philgreenyouilabs** [comment]
Would https://github.com/YOU-i-Labs/appium-youiengine-driver/blob/master/README.md be the appropriate place for that explanation or something in this repo?

**Comment by jlipps** [comment]
I would say dropping a file in here: https://github.com/appium/appium/tree/master/docs/en/appium-setup, and then taking a look at the introductory documentation more generally and adding a reference to Youi automation if you see a good spot. The latter is unnecessary but might raise the visibility.

That being said, it would probably also be a good idea to _also_ have setup instructions in the youiengine driver's README; but I wouldn't want to point users there, I'd want to point developers there. The commands list, for example, is for people importing your driver for development purposes, not for actual users, who will need to use commands from their appropriate client library.

**Pull request merged**